libplatsupport/src/plat/pc99/hpet.c

```diff
@@ -100,30 +100,33 @@ enum {
 
 static inline uint64_t *hpet_get_general_config(void *vaddr)
```

Comment:
I'd say this return type needs the `volatile` qualifier actually.

Reply:
This driver already uses memory fences to ensure correct ordering of memory operations and so adding volatile to all of the memory accesses shouldn't be necessary.

Reply:
but isn't this supposed to be returning a pointer to a 64 device register and not a plain memory location?

Reply:
Yea. When this part of the code was last refactored (https://github.com/seL4/util_libs/commit/478bb400e69df488285907ed07bdfcc44699a3df) all of the volatiles were removed with the assumption that the non-volatile access + memory fences was better because it still allowed the compiler to optimize the code in ways that doesn't introduce bugs.  Unfortunately, now it appears that the compiler is optimizing 32-bit reads and writes into 8-bit in some cases which is potentially against the HPET spec (it says it only supports 32 + 64bit memory operations).  In this case, the memory accesses would need to be forced to stay at 32bit which I believe would require adding inline `asm` accessor functions.  Because that would require rewriting a large part of the driver and a larger amount of work, I opted to using volatile in the one place where the compiler seemed to be performing the bad optimization in a non-critical part of the startup.  I'm not doing any work on this architecture at the moment and I was only helping @lsf37 track down a bug as part of his goal to update the docker containers.

Reply:
I'm happy with it as it as for now. @axel-h if you think this should be refactored in a larger way, we could open an issue for that.